### PR TITLE
Add Dockerfile to simplify build process

### PR DESCRIPTION
**Pull request by ruffsl**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+################################################################################
+FROM builder AS dancer
+
+# Copy entire build directory
+# RUN mkdir /dancer && \
+#     cp -rT build /dancer
+
+# Copy only the lib and bin directories
+RUN mkdir /dancer && \
+    cp build/RGLGazeboPlugin/RGLServerPlugin/*.so /dancer/ && \
```
Comment (ruffsl) [inline Dockerfile:50]: From some reason, it doesn't seem like the `libRobotecGPULidar.so` is being properly installed into the colcon install folder, let alone the build folder.

- https://github.com/RobotecAI/RGLGazeboPlugin/blob/01d26ebf63507aa8ae9c52f6a226eece3f4409a8/RGLServerPlugin/CMakeLists.txt#L73-L75

@msz-rai do you know why this CMake project is writing to the src folder? That should be considered a read only directory.

Comment (msz-rai) [reply Dockerfile:50]: > do you know why this CMake project is writing to the src folder? That should be considered a read only directory.

Good point. At the very early development stage of this Plugin, the Library files were downloaded into src and that's how it stayed. This is not a great practice. I have separated it in [this PR](https://github.com/RobotecAI/RGLGazeboPlugin/pull/42). You are welcome to review it at your convenience.

> From some reason, it doesn't seem like the libRobotecGPULidar.so is being properly installed into the colcon install folder, let alone the build folder.

Could you expand on your issue? Do you mean the way we are installing is incorrect or the library is not present in the directory you expect it to be? I have made a quick test running `colcon build` on my PC and the library was installed into `{PluginDir}/install/RGLServerPlugin/libRobotecGPULidar.so` successfully.

**Comment by ruffsl** [comment]
I've added a few more commits to demonstrate the issues with the existing CMake.

<details>

```shell
> docker build \
    --target=exporter \
    --output=build .
[+] Building 6.3s (12/14)                                                                                                                                  docker:default
 => [internal] load build definition from Dockerfile                                                                                                                 0.0s
 => => transferring dockerfile: 2.11kB                                                                                                                               0.0s
 => [internal] load metadata for docker.io/library/ros:jazzy                                                                                                         1.5s
 => [internal] load .dockerignore                                                                                                                                    0.0s
 => => transferring context: 210B                                                                                                                                    0.0s
 => [base 1/2] FROM docker.io/library/ros:jazzy@sha256:ee4862fe098e18eba8fb0e9843d57241e404aea12bcba57ed037827f16d98ba6                                              0.0s
 => [internal] load build context                                                                                                                                    0.0s
 => => transferring context: 1.53kB                                                                                                                                  0.0s
 => CACHED [base 2/2] RUN mv /etc/apt/apt.conf.d/docker-clean /etc/apt/ &&     echo 'Binary::apt::APT::Keep-Downloaded-Packages "true";'       > /etc/apt/apt.conf.  0.0s
 => CACHED [prepper 1/3] WORKDIR /opt/rgl                                                                                                                            0.0s
 => CACHED [prepper 2/3] COPY package.xml src/RGLGazeboPlugin/package.xml                                                                                            0.0s
 => CACHED [prepper 3/3] RUN --mount=type=cache,sharing=locked,target=/var/cache/apt     rosdep install -y         --from-paths src         --ignore-src             0.0s
 => CACHED [builder 1/3] COPY . /tmp/RGLGazeboPlugin                                                                                                                 0.0s
 => CACHED [builder 2/3] RUN --mount=type=cache,target=src/,rw     cp -r /tmp/RGLGazeboPlugin src/RGLGazeboPlugin                                                    0.0s
 => ERROR [builder 3/3] RUN --mount=type=cache,target=src/,ro     . /opt/ros/jazzy/setup.sh &&     colcon build                                                      4.6s
------                                                                                                                                                                    
 > [builder 3/3] RUN --mount=type=cache,target=src/,ro     . /opt/ros/jazzy/setup.sh &&     colcon build:                                                                 
0.882 WARNING: Package name "RGLGazeboPlugin" does not follow the naming conventions. It should start with a lower case letter and only contain lower case letters, digits, underscores, and dashes.                                                                                                                                                
0.918 Starting >>> RGLGazeboPlugin                                                                                                                                        
4.342 --- stderr: RGLGazeboPlugin                                                                                                                                         
4.342 CMake Error at RGLServerPlugin/CMakeLists.txt:47 (file):
4.342   file DOWNLOAD cannot open file for write.
4.342 
4.342 
4.342 ---
4.342 Failed   <<< RGLGazeboPlugin [3.42s, exited with code 1]
4.400 
4.400 Summary: 0 packages finished [3.60s]
4.400   1 package failed: RGLGazeboPlugin
4.400   1 package had stderr output: RGLGazeboPlugin
------
Dockerfile:38
--------------------
  37 |     
  38 | >>> RUN --mount=type=cache,target=src/,ro \
  39 | >>>     . /opt/ros/$ROS_DISTRO/setup.sh && \
  40 | >>>     colcon build
  41 |     
--------------------
ERROR: failed to solve: process "/bin/sh -c . /opt/ros/$ROS_DISTRO/setup.sh &&     colcon build" did not complete successfully: exit code: 1
```

</details>

Note that reverting 3f69342 will produce the expected built output, but violates the expected CMake/colcon norms.

File: Dockerfile (hunk 1 above)
Comment (ruffsl) [reply Dockerfile]: > Could you expand on your issue?

Please see https://github.com/RobotecAI/RGLGazeboPlugin/pull/40#issuecomment-2226038218 as a concrete test case.

Comment (msz-rai) [reply Dockerfile]: @ruffsl
Thank you for the explanation.

I have added to [my PR](https://github.com/RobotecAI/RGLGazeboPlugin/pull/42) two commits that:
- Remove `CMAKE_INSTALL_PREFIX` hardcoding
- Change the download location of `RobotecGPULidar` project from the source directory to the binary directory.

These modifications required changing the way how to use the custom build of RobotecGPULidar. [This section of the README](https://github.com/RobotecAI/RGLGazeboPlugin/tree/fix/separate-external-project?tab=readme-ov-file#using-custom-build-of-robotecgpulidar) may be useful for your integration.

Please test if my proposed changes resolve your issue. I have tested it (rebased your branch onto my branch) and it worked well for me. Note: One change is required after rebasing - `!/external` needs to be added to the `.dockerignore` file.

Comment (ruffsl) [reply Dockerfile]: Awesome @msz-rai , thanks! I tried by merging your PR with the above changes. Looks to be working, so I've simplified the Dockerfile to reflect that.

**Review by msz-rai — APPROVED** [review]
LGTM!
Thank you for your contribution.
Since https://github.com/RobotecAI/RGLGazeboPlugin/pull/42 has been merged, please change the base branch to `main`. I will merge it then.

**Comment by ruffsl** [comment]
> please change the base branch to main.
Done!

**Pull request merged**